expression/expressions/cmp_subquery.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expressions
+
+import (
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/util/types"
+)
+
+// CompareSubQuery is the expression for "expr cmp (select ...)".
+// See: https://dev.mysql.com/doc/refman/5.7/en/comparisons-using-subqueries.html
+// See: https://dev.mysql.com/doc/refman/5.7/en/any-in-some-subqueries.html
+// See: https://dev.mysql.com/doc/refman/5.7/en/all-subqueries.html
+type CompareSubQuery struct {
+	// L is the left expression
+	L expression.Expression
+	// Op is the comparison opcode.
+	Op opcode.Op
+	// R is the sub query for right expression.
+	R *SubQuery
+	// All is true, we should compare all records in subquery.
+	All bool
+}
+
+// Clone implements the Expression Clone interface.
+func (cs *CompareSubQuery) Clone() (expression.Expression, error) {
+	l, err := cs.L.Clone()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	r, err := cs.R.Clone()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &CompareSubQuery{L: l, Op: cs.Op, R: r.(*SubQuery), All: cs.All}, nil
+}
+
+// IsStatic implements the Expression IsStatic interface.
+func (cs *CompareSubQuery) IsStatic() bool {
+	return cs.L.IsStatic() && cs.R.IsStatic()
+}
+
+// String implements the Expression String interface.
+func (cs *CompareSubQuery) String() string {
+	anyOrAll := "any"
+	if cs.All {
+		anyOrAll = "all"
+	}
+
+	return fmt.Sprintf("%s %s %s %s", cs.L, cs.Op, anyOrAll, cs.R)
+}
+
+// Eval implements the Expression Eval interface.
+func (cs *CompareSubQuery) Eval(ctx context.Context, args map[interface{}]interface{}) (interface{}, error) {
+	if err := hasSameColumnCount(ctx, cs.L, cs.R); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	in, err := cs.L.Eval(ctx, args)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if in == nil {
+		return nil, nil
+	}
+
+	p, err := cs.R.Plan(ctx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	res := []interface{}{}
```

Comment:
hold the sub select values for next check, e.g.

```
select * from t where c1 > (select c2 from t2); 
```

if not, we will run `select c2 from t2` many times.